### PR TITLE
feat: improve error messages on API failures

### LINKED ISSUE #39: 🐞 [Bug Report] - Product Item not loading on hard refresh
**Describe the bug**
On the client side, when attempting a hard refresh on a product item `/product/3`, the page reveals a "Not found" error.

**To Reproduce**
1. Go to local version or deployed version of avocano and route to a product item page that's not on root. (i.e `/product/3`)
2. Click on refresh button on browser.

**Expected behavior**
The product item page should refresh as usual with its content.

**Screenshots**
Screenshot 2023-02-08 at 1 29 39 PM
Screenshot 2023-02-08 at 1 29 29 PM

**Additional context**
N/A

### LINKED ISSUE #205: 🐞 [Bug Report] - Querystring without value assignment breaks page routing
**Describe the bug**

Avocano is stuck on the loading page when the URL includes a querystring such as `?cachebuster` instead of `?cachebuster=12345`.

This is shown in the browser console as
Screenshot 2023-04-26 at 10 50 26 AM

There are no XHR requests for the site_config from django.

**Expected behavior**
Unknown querystring parameters are mostly ignored or passed thru where it makes sense.

**Additional context**

This problem came up at the same time that I encountered a stubborn 404 page after terraform apply from the other repo: https://github.com/GoogleCloudPlatform/terraform-dynamic-python-webapp/issues/57

### LINKED ISSUE #230: 💡 [Feature Request]: Infrastructure issue resilience, useful failure states
Feature request: have the application be more resilient to infrastructure errors

See https://github.com/GoogleCloudPlatform/terraform-dynamic-python-webapp/issues/64#issuecomment-1544776564

The server, once deployed, should be resilient to lack of data and database: 

* /active/ endpoints with no active elements should return purposeful error codes in a format the client can react to. Example:
no active product?
  - OLD: 500 (command line: store.models.Product.DoesNotExist)
  - NEW: 404 with message (or equiv)

* The client should be resilient to errors from the server, and should have some sort of interface showing there's a problem. 

  - front end should react to codes from server, and display either the server message, or something to the effect of "something went wrong"

* The client should be resilient to non-connectivity to the server.
  - front end should stop the 'loading' page, and show an error page in the event the API GET is refused

### LINKED ISSUE #236: 🐞 [Bug Report] - Refreshing a valid URL through the browser results in a 404
**Describe the bug**
When refreshing Avocano on a page that is not root (for example, `/products`), the page shows a 404 asking to manually click a link to get back to the site.

**To Reproduce**
1. Visit the `/products` page
2. Refresh the page through the web browser (cmd+r)

**Expected behavior**
I expected the web app to load the expected page (in this case, the products page)

**Screenshots**
image

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by glasnt**
### Resolves #230

Failure states: 

No local database, page returns error screen
Screenshot 2023-06-22 at 1 11 15 pm

No primed database, page returns error screen
Screenshot 2023-06-22 at 1 11 53 pm

Page will optionally use the response error page from Django's debug mode to pull out the reason, saving users the click. 

No active products, "No active products" displayed as an error message
Screenshot 2023-06-22 at 1 13 26 pm

No products at all, /products displays such.

Screenshot 2023-06-22 at 1 13 52 pm

Server uses `get_object_or_404` to serve useful errors, and client API now parses those errors into either a standalone error page, or error messages inline. 

'Ghost' errors should now be prevented (where the default placeholder image is a "ghost", as opposed to these much more useful and less haunted error screens)

### Resolves #236

Unable to reproduce changes in current deployment, may have been already fixed, and now resolved. 

### Resolves #39

Vendors lit-element-router, applies fix from unmerged PR

### Resolves #205

With vendored lit-element-router, fixes querystring parsing to RFC3986 

### Cleanup fetch.js

Converts each individual fetch method into two general methods for GET and POST, handling common error elevation with GETs, doing common CSRF tasks in POSTs

### Development fix

Ensures localhost for both local client and server are allowed in CSRF trusted domains, allowing for local client dev POST testing, and server POST testing.

### CODE REVIEW

**Comment by glasnt** [comment]
header-check: the vendored code is [ISC](https://github.com/hamedasemi/lit-element-router/blob/v2.0.3/package.json#L35), pending internal discussions about best practices for headers in this instance.

**Review by pattishin — COMMENTED** [review]
👋🏼 Nice! I mostly added nit comments around organizing & cleaning up. 🧹

File: client/src/utils/fetch.js [hunk 1]
```diff
@@ -18,209 +18,123 @@ const baseRequest = {
...
 };
 
-/**
- * getProduct()
- *
- * Retrieves product at specified id defined from django api
- * GET /product/{productId}
- */
-export const getProduct = async productId => {
+async function _getAPI(uri) {
```
[the remark this answers is not in the document]
Comment (glasnt) [reply client/src/utils/fetch.js:22]: I think this is fine as it is for now, unless there's a speed improvement that can be made.

**Review by pattishin — APPROVED** [review]
Just needs some header lint massaging, but lgtm - ty katie!
This will also resolve https://github.com/GoogleCloudPlatform/avocano/issues/161 `querySelectorAll` issue.

**Comment by glasnt** [comment]
Following up on the linting issue: the config needed to be in "header-check**er**-lint.yml" to be picked up. If this is correct, you _can_ do PR-based updates.

**Pull request merged**